README.rst
====================================
Scrapy & Autoextract API integration
====================================

.. image:: https://img.shields.io/pypi/v/scrapy-autoextract.svg
   :target: https://pypi.org/project/scrapy-autoextract/
   :alt: PyPI Version

.. image:: https://img.shields.io/pypi/pyversions/scrapy-autoextract.svg
    :target: https://pypi.org/project/scrapy-autoextract/
    :alt: Supported Python Versions

.. image:: https://github.com/scrapinghub/scrapy-autoextract/workflows/tox/badge.svg
   :target: https://github.com/scrapinghub/scrapy-autoextract/actions
   :alt: Build Status

.. image:: https://codecov.io/gh/scrapinghub/scrapy-autoextract/branch/master/graph/badge.svg?token=D6DQUSkios
    :target: https://codecov.io/gh/scrapinghub/scrapy-autoextract
    :alt: Coverage report


This library integrates Zyte's AI Enabled Automatic Data Extraction
into a Scrapy spider by two different means:

* with a downloader middleware that injects the AutoExtract responses into ``response.meta['autoextract']``
  for consumption by the spider.
* with a `scrapy-poet`_ provider that injects the responses as callback parameters.


Installation
============

::

    pip install scrapy-autoextract

scrapy-autoextract requires Python 3.7+ for the download middleware and Python 3.7+ for the scrapy-poet provider


Usage
=====

There are two different ways to consume the AutoExtract API with this library:

* using our Scrapy middleware
* using our Page Object provider

The middleware
--------------

The middleware is opt-in and can be explicitly enabled per request,
with the ``{'autoextract': {'enabled': True}}`` request meta.
All the options below can be set either in the project settings file,
or just for specific spiders, in the ``custom_settings`` dict.

Within the spider, consuming the AutoExtract result is as easy as::

    def parse(self, response):
        yield response.meta['autoextract']

Configuration
^^^^^^^^^^^^^

Add the AutoExtract downloader middleware in the settings file::

    DOWNLOADER_MIDDLEWARES = {
        'scrapy_autoextract.AutoExtractMiddleware': 543,
    }

Note that this should be the last downloader middleware to be executed.

The providers
-------------

Another way of consuming AutoExtract API is using the Page Objects pattern
proposed by the `web-poet`_ library and implemented by `scrapy-poet`_.

Items returned by Page Objects are defined in the `autoextract-poet`_
library.

Within the spider, consuming the AutoExtract result is as easy as::

    import scrapy
    from autoextract_poet.pages import AutoExtractArticlePage

    class SampleSpider(scrapy.Spider):
        name = "sample"

        def parse(self, response, article_page: AutoExtractArticlePage):
            # We're making two requests here:
            # - one through Scrapy to build the response argument
            # - the other through the providers to build the article_page argument
            yield article_page.to_item()

Note that on the example above, we're going to perform two requests:

* one goes through Scrapy (it might use Smart Proxy, Splash or no proxy at all, depending on your configuration)
* another goes through AutoExtract API using `zyte-autoextract`_

If you don't need the additional request going through Scrapy,
you can annotate the response argument of your callback with ``DummyResponse``.
This will ignore the Scrapy request and only the AutoExtract API will be fetched.

For example::

    import scrapy
    from autoextract_poet.pages import AutoExtractArticlePage
    from scrapy_poet import DummyResponse

    class SampleSpider(scrapy.Spider):
        name = "sample"

        def parse(self, response: DummyResponse, article_page: AutoExtractArticlePage):
            # We're making a single request here to build the article argument
            yield article_page.to_item()


The examples above extract an article from the page, but you may want to
extract a different type of item, like a product or a job posting. It is
as easy as using the correct type annotation in the callback. This
is how the callback looks like if we need to extract a real state
from the page::

    def parse(self,
              response: DummyResponse,
              real_estate_page: AutoExtractRealEstatePage):
        yield real_estate_page.to_item()

You can even use ``AutoExtractWebPage`` if what you need is the raw browser HTML to
extract some additional data. Visit the full list of `supported page types
<https://docs.zyte.com/automatic-extraction.html#result-fields>`_ to get a better idea
of the supported pages.

Lastly, if you have a an AutoExtract subscription with `fullHtml` set to True,
you can access the HTML data that was used by AutoExtract in case you need it.
Here's an example:

.. code-block:: python

    def parse_product(self, response: DummyResponse, product_page: AutoExtractProductPage, html_page: AutoExtractWebPage):
        product_item = product_page.to_item()

        # You can easily interact with the html_page using these selectors.
        html_page.css(...)
        html_page.xpath(...)

Configuration
^^^^^^^^^^^^^

First, you need to configure scrapy-poet as described on `scrapy-poet's documentation`_
and then enable AutoExtract providers by putting the following code to Scrapy's ``settings.py`` file::

    # Install AutoExtract provider
    SCRAPY_POET_PROVIDERS = {"scrapy_autoextract.AutoExtractProvider": 500}

    # Enable scrapy-poet's provider injection middleware
    DOWNLOADER_MIDDLEWARES = {
        'scrapy_poet.InjectionMiddleware': 543,
    }

    # Configure Twisted's reactor for asyncio support on Scrapy
    TWISTED_REACTOR = 'twisted.internet.asyncioreactor.AsyncioSelectorReactor'

Currently, our providers are implemented using asyncio.
Scrapy has introduced asyncio support since version 2.0
but as of Scrapy 2.3 you need to manually enable it by configuring Twisted's default reactor.
Check `Scrapy's asyncio documentation`_ for more information.

Checklist:

* scrapy-poet is installed and downloader/injector middleware is configured
* autoextract-poet is installed (page inputs are imported from this lib)
* providers are configured on settings.py
* Scrapy's asyncio support is enabled on settings.py

Now you should be ready to use our AutoExtract providers.

Exceptions
^^^^^^^^^^

While trying to fetch AutoExtract API, providers might raise some exceptions.
Those exceptions might come from scrapy-autoextract providers themselves,
`zyte-autoextract`_, or by other means (e.g. ``ConnectionError``).
For example:

* ``autoextract.aio.errors.RequestError``: raised when a `Request-level error`_ is returned
* ``scrapy_autoextract.errors.QueryError``: raised when a `Query-level error`_ is returned

Check `zyte-autoextract's async errors`_ for other exception definitions.

You can capture those exceptions using an error callback (``errback``)::

    import scrapy
    from autoextract.aio.errors import RequestError
    from autoextract_poet.pages import AutoExtractArticlePage
    from scrapy_autoextract.errors import QueryError
    from scrapy_poet import DummyResponse
    from twisted.python.failure import Failure

    class SampleSpider(scrapy.Spider):
        name = "sample"
        urls = [...]

        def start_requests(self):
            for url in self.urls:
                yield scrapy.Request(url, callback=self.parse_article,
                                     errback=self.errback_article)

        def parse_article(self, response: DummyResponse,
                          article_page: AutoExtractArticlePage):
            yield article_page.to_item()

        def errback_article(self, failure: Failure):
            if failure.check(RequestError):
                self.logger.error(f"RequestError on {failure.request.url}")

            if failure.check(QueryError):
                self.logger.error(f"QueryError: {failure.value.message}")

See `Scrapy documentation <https://docs.scrapy.org/en/latest/topics/request-response.html#using-errbacks-to-catch-exceptions-in-request-processing>`_
for more details on how to capture exceptions using request's errback.

Settings
========

Middleware settings
-------------------

- ``AUTOEXTRACT_USER`` [mandatory] your AutoExtract API key.
- ``AUTOEXTRACT_URL`` [optional] the AutoExtract service url. Defaults to autoextract.scrapinghub.com.
- ``AUTOEXTRACT_TIMEOUT`` [optional] sets the response timeout from AutoExtract. Defaults to 660 seconds.
  Can also be defined by setting the "download_timeout" in the request.meta.
- ``AUTOEXTRACT_PAGE_TYPE`` [mandatory] defines the kind of document to be extracted.
  See currently `supported page types <https://docs.zyte.com/automatic-extraction.html#result-fields>`_.
  Can also be defined on ``spider.page_type``, or ``{'autoextract': {'pageType': '...'}}`` request meta.
  This is required for the AutoExtract classifier to know what the extraction result should be (article, job posting, product, etc.).
- `extra` [optional] allows sending extra payload data to your AutoExtract request.
  Must be specified as ``{'autoextract': {'extra': {}}}`` request meta and must be a dict.
- ``AUTOEXTRACT_SLOT_POLICY`` [optional] Download concurrency options. Defaults to ``SlotPolicy.PER_DOMAIN``
  - If set to ``SlotPolicy.PER_DOMAIN``, then consider setting ``SCHEDULER_PRIORITY_QUEUE = 'scrapy.pqueues.DownloaderAwarePriorityQueue'``
  to make better usage of AutoExtract concurrency and avoid delays.
- ``AUTOEXTRACT_RESPONSE_ERROR_LOG_LEVEL`` [optional] Sets the logging level for logging the errors encountered.
  _(default: logging.DEBUG)_
- ``AUTOEXTRACT_ALLOWED_RESPONSE_ERRORS`` [optional] A collection of error strings from AutoExtract's responses to
  prevent ``AutoExtractError`` from being raised. The collection of strings defined here will be combined with the default values.
  _(default: ``{"Downloader error: http404"}`` that prevents valid 404 website responses from erroring out)_

Provider settings
-----------------

- ``AUTOEXTRACT_USER`` [optional] is your AutoExtract API key. If not set, it is
  taken from ZYTE_AUTOEXTRACT_KEY environment variable.
- ``AUTOEXTRACT_URL`` [optional] the AutoExtract service url.
  Defaults to the official AutoExtract endpoint.
- ``AUTOEXTRACT_MAX_QUERY_ERROR_RETRIES`` [optional] Max number of retries for
  Query-level errors. Defaults to ``0``.
- ``AUTOEXTRACT_CONCURRENT_REQUESTS_PER_DOMAIN`` [optional] Max number
  of concurrent requests per domain. If not set, the provider will search
  for the `CONCURRENT_REQUESTS_PER_DOMAIN` (defaults to ``8``) setting instead.
- ``AUTOEXTRACT_CACHE_FILENAME`` [optional] Filename of a .sqlite file that will
  be placed in the ``.scrapy`` folder. File will be created if it doesn't exist.
  Cache is useful for development; AutoExtract requests bypass standard Scrapy
  cache when providers are used.
- ``AUTOEXTRACT_CACHE_GZIP`` [optional] when True (default), cached AutoExtract
  responses are compressed using gzip. Set this option to False to turn
  compression off.

Limitations
===========

When using the AutoExtract middleware, there are some limitations.

* The incoming spider request is rendered by AutoExtract, not just downloaded by Scrapy,
  which can change the result - the IP is different, headers are different, etc.
* Only GET requests are supported
* Custom headers and cookies are not supported (i.e. Scrapy features to set them don't work)
* Proxies are not supported (they would work incorrectly,
  sitting between Scrapy and AutoExtract, instead of AutoExtract and website)
* AutoThrottle extension can work incorrectly for AutoExtract requests,
  because AutoExtract timing can be much larger than time required to download a page,
  so it's best to use ``AUTHTHROTTLE_ENABLED=False`` in the settings.
* Redirects are handled by AutoExtract, not by Scrapy,
  so these kinds of middlewares might have no effect
* 429 errors could be handled as standard retries when using Scrapy middleware,
  but they're handled properly and automatically with scrapy-poet integration,
  as it relies on `zyte-autoextract`_.
  You may lose some responses with the middleware approach.
* Overall, retries have a better behavior with scrapy-poet integration
  and it includes support for automatic Query-level errors retries with
  no need to change ``RETRY_HTTP_CODES``.
* AutoExtract-specific cache (``AUTOEXTRACT_CACHE_FILENAME``) is not supported

When using the AutoExtract providers, be aware that:

* With scrapy-poet integration, retry requests don't go through Scrapy

.. _`web-poet`: https://github.com/scrapinghub/web-poet
.. _`scrapy-poet`: https://github.com/scrapinghub/scrapy-poet
.. _`autoextract-poet`: https://github.com/scrapinghub/autoextract-poet
.. _`zyte-autoextract`: https://github.com/zytedata/zyte-autoextract
.. _`zyte-autoextract's async errors`: https://github.com/zytedata/zyte-autoextract/blob/master/autoextract/aio/errors.py
.. _`scrapy-poet's documentation`: https://scrapy-poet.readthedocs.io/en/latest/intro/tutorial.html#configuring-the-project
.. _`Scrapy's asyncio documentation`: https://docs.scrapy.org/en/latest/topics/asyncio.html
.. _`Request-level error`: https://doc.scrapinghub.com/autoextract.html#request-level
.. _`Query-level error`: https://doc.scrapinghub.com/autoextract.html#query-level
.. _`supported page types`: https://autoextract-poet.readthedocs.io/en/stable/_autosummary/autoextract_poet.pages.html#module-autoextract_poet.pages


CHANGES.rst
Changes
=======

TBR
------------------

* removed support for Python 3.6
* added support for Python 3.10
* add _optional_ setting ``AUTOEXTRACT_RESPONSE_ERROR_LOG_LEVEL`` to set the
  logging level for the errors encountered.
* add _optional_ setting ``AUTOEXTRACT_ALLOWED_RESPONSE_ERRORS`` which is a
  collection of error strings from AutoExtract's responses to prevent
  ``AutoExtractError`` from being raised when errors are valid _(e.g. a website
  returns a valid page but with a 404 response status)_.

0.7.0 (2021-08-05)
------------------

* Support for all Automatic Extraction API page types by upgrading to
  ``autoextract-poet`` 0.3.0
* Rename Scrapinghub references to Zyte
* Update README

0.6.1 (2021-06-02)
------------------

* A cache of AutoExtract requests can now be enabled by
  setting ``AUTOEXTRACT_CACHE_FILENAME``.

0.6.0 (2021-06-01)
------------------
* Switch from ``scrapinghub-autoextract`` to ``zyte-autoextract`` package,
  following its rename.
* Upgrade ``scrapy-poet`` to 0.2.0+.

Note that the switch is backwards incompatible if you're
relying on ``SCRAPINHUB_AUTOEXTRACT_KEY`` environment variable
to set an API key; it is no longer working. Please use
either ``ZYTE_ATOEXTRACT_KEY`` env variable or ``AUTOEXTRACT_USER``
setting instead.

0.5.2 (2021-01-27)
------------------

* Upgrade ``autoextract-poet`` to 0.2.1
* Upgrade ``scrapinghub-autoextract`` to 0.6.1

0.5.1 (2021-01-22)
------------------
* AUTOEXTRACT_MAX_QUERY_ERROR_RETRIES default value is set to 0, to account
  for backend changes in AutoExtract API.

0.5.0 (2021-01-21)
------------------

* Mayor internal and API refactor. No backwards compatible.
* Support for the new ``autoextract-poet`` 0.2.0 types.
* Use of the new providers interface introduced in ``scrapy-poet``  0.1.0
* ``scrapinghub-autoextract`` AutoExtract client updated to 0.6.0
* CI is moved from Travis to Github Actions.
* Python 3.9 support.


scrapy_autoextract/middlewares.py
import time
import json
import logging

import scrapy
from scrapy import signals
from scrapy.http import Headers, HtmlResponse
from scrapy.utils.python import global_object_name
from scrapy.exceptions import IgnoreRequest, DropItem
from w3lib.http import basic_auth_header

from .__version__ import __version__

logger = logging.getLogger(__name__)

AUTOEXTRACT_META_KEY = '_autoextract_processed'
USER_AGENT = 'scrapy-autoextract/{} scrapy/{}'.format(__version__, scrapy.__version__)
MAX_ERROR_BODY = 2000


class AutoExtractError(Exception):
    pass


class AutoExtractConfigError(Exception):
    pass


class SlotPolicy(object):
    PER_DOMAIN = 'per_domain'
    SINGLE_SLOT = 'single_slot'
    SCRAPY_DEFAULT = 'scrapy_default'


class AutoExtractMiddleware(object):
    """
    Middleware that allows a Scrapy Spider to receive Automatic Data Extraction
    results within the `response.meta`.

    The service URL can be specified with `AUTOEXTRACT_URL` in scrapy settings.
    """

    DEFAULT_URL = 'https://autoextract.scrapinghub.com/v1/extract'
    DEFAULT_TIMEOUT = 660
    DEFAULT_SLOT_POLICY = SlotPolicy.PER_DOMAIN
    DEFAULT_ALLOWED_REPONSE_ERRORS = {
        "Downloader error: http404",  # some sites return 404 as a valid response
    }

    def __init__(self, crawler):
        self.crawler = crawler
        self.settings = crawler.settings
        self._api_user = self.settings['AUTOEXTRACT_USER']
        self._api_pass = ''
        self.page_type = self.settings['AUTOEXTRACT_PAGE_TYPE']
        self._log_response_error_level = self.settings.get(
            "AUTOEXTRACT_RESPONSE_ERROR_LOG_LEVEL", logging.DEBUG
        )
        self._allowed_response_errors = (
            set(self.settings.get("AUTOEXTRACT_ALLOWED_RESPONSE_ERRORS", []))
            | self.DEFAULT_ALLOWED_REPONSE_ERRORS
        )
        if not self.page_type:
            self.page_type = getattr(crawler.spider, 'page_type', None)
        self.timeout = max(
            self.settings.getint('AUTOEXTRACT_TIMEOUT', self.DEFAULT_TIMEOUT),
            self.settings.getint('DOWNLOAD_TIMEOUT', 180),
        )
        self.slot_policy = self.settings.get('AUTOEXTRACT_SLOT_POLICY', self.DEFAULT_SLOT_POLICY)

        self._api_url = self.settings.get('AUTOEXTRACT_URL', self.DEFAULT_URL)
        logger.info('Using AutoExtract API URL: %s', self._api_url, extra={'spider': crawler.spider})

        self.nr_resp = 0
        self.avg_latency = 0
        self.max_latency = 0
        self.total_latency = 0

    @classmethod
    def from_crawler(cls, crawler):
        if not crawler.settings.get('AUTOEXTRACT_USER'):
            raise AutoExtractConfigError('AUTOEXTRACT_USER is required')
        o = cls(crawler)
        crawler.signals.connect(o.spider_closed, signal=signals.spider_closed)
        return o

    def spider_closed(self, spider):
        if self.nr_resp > 0:
            self.autoextract_latency_stats()
            logger.info(
                'Total "%i" responses from AutoExtract, average latency=%.3f, max latency=%.3f',
                self.nr_resp,
                self.avg_latency,
                self.max_latency,
                extra={'spider': spider},
            )

    def process_request(self, request, spider):
        """
        The request will be passed to the AutoExtract server only if the request
        is explicitly enabled with `{'autoextract': {'enabled': True}}` meta.
        The page type value must be also present, either in the
        AUTOEXTRACT_PAGE_TYPE option, or in `{'autoextract': {'pageType': '...'}}` meta.
        """
        if not self._is_enabled_for_request(request):
            return

        # If the request was already processed by AutoExtract
        if request.meta.get(AUTOEXTRACT_META_KEY):
            return

        if request.method != 'GET':
            raise AutoExtractError('Only GET requests are supported by AutoExtract')

        request.meta[AUTOEXTRACT_META_KEY] = {
            'original_url': request.url,
            'timing': {
                'start_ts': time.time()
            },
        }

        # Maybe over-write the page type value from the request
        page_type = self._check_page_type(request)
        logger.debug('Process AutoExtract request for %s URL %s',
                     page_type,
                     request,
                     extra={'spider': spider})

        # Define request timeout
        request.meta['download_timeout'] = self.timeout

        # Define concurrency settings
        self._set_download_slot(request, request.meta)

        payload = {'url': request.url, 'pageType': page_type}

        # Add the extra payload, if available
        extra_payload = self._get_meta_name(request, 'extra')
        if extra_payload:
            payload.update(extra_payload)

        headers = Headers({
            'Content-Type': 'application/json',
            'User-Agent': USER_AGENT,
            'Authorization': basic_auth_header(self._api_user, self._api_pass)
        })
        # Update the headers, if provided
        extra_headers = self._get_meta_name(request, 'headers')
        if extra_headers:
            headers.update(extra_headers)

        new_request = request.replace(
            url=self._api_url,
            method='POST',
            headers=headers,
            body=json.dumps([payload], sort_keys=True),
        )

        self.inc_metric('autoextract/request_count')
        return new_request

    def process_response(self, request, response, spider):
        if not self._is_enabled_for_request(request):
            return response

        # If the request was never processed by AutoExtract
        if not request.meta.get(AUTOEXTRACT_META_KEY):
            return response

        url = request.meta[AUTOEXTRACT_META_KEY]['original_url']
        body = response.body.decode('utf8')

        try:
            response_object = json.loads(body)
        except Exception:
            self.inc_metric('autoextract/errors/json_decode')
            self._log_response_error(response, body)
            raise AutoExtractError('Cannot parse JSON response from AutoExtract'
                                   ' for {}: {}'.format(url, response.body[:MAX_ERROR_BODY]))

        if response.status != 200:
            self.inc_metric('autoextract/errors/response_error/{}'.format(response.status))
            self._log_response_error(response, body)
            raise AutoExtractError('Received error from AutoExtract for '
                                   '{}: {}'.format(url, response_object))

        result = None
        if isinstance(response_object, list):
            result = response_object[0]
        else:
            self.inc_metric('autoextract/errors/type_error')
            self._log_response_error(response, body)
            raise AutoExtractError('Received invalid response from AutoExtract for '
                                   '{}: {}'.format(url, response_object))

        error = result.get('error')
        if error and error not in self._allowed_response_errors:
            self.inc_metric('autoextract/errors/result_error')
            self._log_response_error(response, body)
            raise AutoExtractError('Received error from AutoExtract for '
                                   '{}: {}'.format(url, result["error"]))

        stop_time = time.time()
        latency = stop_time - request.meta[AUTOEXTRACT_META_KEY]['timing']['start_ts']

        self.nr_resp += 1
        self.total_latency += latency
        self.max_latency = max(self.max_latency, latency)
        self.avg_latency = float(self.total_latency) / self.nr_resp
        self.autoextract_latency_stats()

        autoextract = request.meta.pop(AUTOEXTRACT_META_KEY)
        # Middleware-level timings
        autoextract['timing'].update({'end_ts': stop_time, 'latency': latency})

        page_type = self._check_page_type(request)
        logger.debug('AutoExtract latency for %s URL %s was %.3fs',
                     page_type,
                     url,
                     latency,
                     extra={'spider': spider})

        # The AutoExtract processed item is added here
        autoextract[page_type] = result.get(page_type) or {}
        request.meta['autoextract'] = autoextract
        page_html = result.pop('html', '<body></body>')

        return HtmlResponse(
            url,
            request=request,
            body=page_html.encode('utf-8'),
            encoding='utf-8',
        )

    def process_exception(self, request, exception, spider):
        if isinstance(exception, (IgnoreRequest, DropItem)):
            return
        if not self._is_enabled_for_request(request):
            return

        autoextract = request.meta.pop(AUTOEXTRACT_META_KEY)
        stop_time = time.time()
        latency = time.time() - autoextract['timing']['start_ts']
        autoextract['timing'].update({'end_ts': stop_time, 'latency': latency})

        # Make sure to log all unknown failures
        logger.warning('AutoExtract failure after %.3fs for %s: %s',
                       latency,
                       autoextract['original_url'],
                       repr(exception),
                       extra={'spider': spider})

        request.meta['autoextract'] = autoextract
        ex_class = global_object_name(exception.__class__)
        self.inc_metric('autoextract/errors/total_count', spider=spider)
        self.inc_metric('autoextract/errors/type_count/%s' % ex_class, spider=spider)

    def _is_enabled_for_request(self, request):
        if 'autoextract' not in request.meta:
            return False
        if not request.meta.get('autoextract', {}).get('enabled'):
            return False
        return True

    def _check_page_type(self, request):
        # Use pageType value from the request.meta['autoextract']
        # and fallback to the value from middleware
        page_type = request.meta.get('autoextract', {}).get('pageType', self.page_type)
        if not page_type or not isinstance(page_type, str):
            raise AutoExtractConfigError('Invalid pageType value: {}'.format(page_type))
        return page_type

    def _set_download_slot(self, request, meta):
        if self.slot_policy == SlotPolicy.PER_DOMAIN:
            # Group requests by domain to respect download
            # delays and concurrency options
            slot = self.crawler.engine.downloader._get_slot_key(request, None)
            meta['download_slot'] = slot
        elif self.slot_policy == SlotPolicy.SINGLE_SLOT:
            # Use a single slot for all AutoExtract requests
            meta['download_slot'] = '__AutoExtract__'
        # Else, use standard Scrapy concurrency setup

    def _get_meta_name(self, request, name):
        extra_data = None
        if request.meta['autoextract'].get(name):
            if not isinstance(request.meta['autoextract'][name], dict):
                raise AutoExtractError('Invalid type for "{}" meta'.format(name))
            extra_data = request.meta['autoextract'][name]
        return extra_data

    def inc_metric(self, key, **kwargs):
        self.crawler.stats.inc_value(key, **kwargs)

    def set_metric(self, key, value):
        self.crawler.stats.set_value(key, value)

    def _log_response_error(self, response, body):
        if len(body) > MAX_ERROR_BODY:
            half_body = MAX_ERROR_BODY // 2
            body = body[:half_body] + ' [...] ' + body[-half_body:]
        logger.log(
            self._log_response_error_level,
            'AutoExtract response status=%i  headers=%s  content=%s',
            response.status,
            response.headers.to_unicode_dict(),
            body
        )

    def autoextract_latency_stats(self):
        self.set_metric('autoextract/response_count', self.nr_resp)
        self.set_metric('autoextract/response_avg_latency', self.avg_latency)
        self.set_metric('autoextract/response_max_latency', self.max_latency)
